Updated user guide + fix to predict_proba + updated tree-based weights and similarity

NiklasPfister opened the pull request:
This is a first iteration of the updated documentation. It is not fully complete but good enough for a first merge.

Additionally to the documentation there are several fixes for problems I came across while writing the user guide and creating examples:
1) There was a bug in predict_proba where it converted the probabilities to integers.
2) The tree-based weights did not fully match what I originally intended. Specifically, it did not account for the repeated observations that can happen when fitting a random forest. Additionally, I switched the scaling to row-wise as this seems more inline with most existing works on this.
3) I removed the scaling from the similarity as I don't think it makes sense (there is now a formal math definition in the user guide).

I also added a tests for each of these points.

svbrodersen commented on pyproject.toml:
```diff
@@ -8,5 +8,3 @@ max-line-length = 127
 max-line-length = 127
 per-file-ignores = ['__init__.py:F401']
 
-[tool.cibuildwheel]
```
Why was this removed? This will make it build for CPython 6,7,8,9, which is not possible as far as I am aware.

NiklasPfister replied:
Sorry it should not be removed. The merge reverted a few things backward and I missed this change. I'll fix it.

svbrodersen commented on src/adaXT/decision_tree/decision_tree.pyx:
```diff
@@ -155,24 +155,28 @@ class DecisionTree(BaseModel):
     def _tree_based_weights(self, hash0: dict, hash1: dict, size_X0: int,
-                            size_X1: int, scale: bool) -> np.ndarray:
+                            size_X1: int, scaling: str) -> np.ndarray:
         matrix = np.zeros((size_X0, size_X1))
```
Given there is now an inner loop, it might a good idea to give the loop a type definition for performance.

NiklasPfister replied:
Yes, good point. Can you maybe propose something here. Also do you see an alternative of getting around the inner loop, I didn't manage to find anything so I stuck with this?

svbrodersen commented on src/adaXT/decision_tree/decision_tree.pyx:
```diff
@@ -155,24 +155,28 @@ class DecisionTree(BaseModel):
         return matrix
 
-    def similarity(self, X0: ArrayLike, X1: ArrayLike, scale: bool = True):
+    def similarity(self, X0: ArrayLike, X1: ArrayLike):
```
We don't want there to be no scaling for similarity any more?

NiklasPfister replied:
No, I couldnt come up with an example why one would want to do this so I thought we might as well just remove it.

svbrodersen commented on src/adaXT/predict/predict.pyi:
```diff
@@ -58,8 +58,8 @@ class PredictClassification(Predict):
         """
-        Predicts the LeafNode every $x \in X$ would fall into and returns the
-        class which occurs the most amount of times within the LeafNode
+        Predicts the LeafNode each row in $X$ would fall into and returns the
```
for each row in X

svbrodersen commented on src/adaXT/predict/predict.pyi:
```diff
@@ -84,13 +84,13 @@ class PredictClassification(Predict):
 class PredictRegression(Predict):
     """
-    Default used for Regression.
+        The default prediction calls for the Regression tree type.
```
The default prediction calls the Regression tree type

NiklasPfister replied:
Updated this entire file

svbrodersen commented on src/adaXT/predict/predict.pyi:
```diff
@@ -84,13 +84,13 @@ class PredictClassification(Predict):
         """
-        Predicts which LeafNode each $x \in X$ would land in, and calculates the
-        mean value of all elements in the LeafNode.
+        Predicts which LeafNode each row in $X$ would land in, and calculates the
```
For each row in X

svbrodersen commented on src/adaXT/predict/predict.pyx:
```diff
@@ -175,7 +175,7 @@ cdef class PredictClassification(Predict):
             if self.classes is not None:
                 ret_val.append(cur_node.value)
-        return np.array(ret_val, dtype=int)
+        return np.array(ret_val)
```
For the classification specifically, do we not want to return an integer array?

NiklasPfister replied:
Yes, but this happens already in `__predict` or not?

svbrodersen replied:
I take this back. It should never just be int. One might have a classification, where the outcome is a float of some kind.

svbrodersen commented on src/adaXT/predict/predict.pyx:
```diff
@@ -196,7 +196,7 @@ cdef class PredictClassification(Predict):
-        predictions = parallel.async_map(default_predict, trees, X=X_new,
+        predictions = parallel.async_map(predict_default, trees, X=X_new,
                                          **kwargs)
         return np.array(np.apply_along_axis(mode, 0, predictions), dtype=DOUBLE)
```
Same with the dtype here. Perhaps we want int for classification.

NiklasPfister replied:
Mhh not sure I follow. Here its actually also the wrong way around. If we have predict_proba=True it should be DOUBLE and if not it should be int. Or am I missing something?

svbrodersen commented on src/adaXT/random_forest/random_forest.py:
```diff
@@ -36,7 +32,7 @@ def tree_based_weights(
         size_X0=size_X0,
         size_X1=size_X1,
-        scale=scale,
+        scaling=scaling,
```
Should the scaling with a string input be outwards facing? Would you not always want row wise scaling when scaling tree_based_weights? Do we maybe want to make it a bool and then convert internally?

NiklasPfister replied:
Its the same function that is also used for the similarity which is again a different scaling so I think it needs to be outward facing here or not?

svbrodersen commented on src/adaXT/random_forest/random_forest.py:
```diff
@@ -481,11 +504,7 @@ def predict_weights(
-            scaling = "symmetric"
-        else:
-            scaling = "none"
+    def similarity(self, X0: np.ndarray, X1: np.ndarray):
```
Same about no scaling similarity

NiklasPfister replied:
See above

svbrodersen reviewed (commented):
Generally fine. Didn't look through markdown files, but the code. Have some questions throughout.

NiklasPfister commented:
Added some more fixes/changes:
- predict_proba for forests was still incorrect (hopefully its correct now)
- There was bug in refit_leaf_nodes: If the tree only consisted of a root node it set it None and did not update the leaf_nodes.
- When trying to integrate the RF weights I needed some additional flexibility on the sampling schemes, so I changed sampling_parameters to sampling_args which is dict now. I feel this is a bit easier to understand than having parameters that are too different.

NiklasPfister commented:
One additional fix: The impurity for LocalLinear and LocalQuadratic was not scaled properly (similar to Squared_error).

svbrodersen commented on src/adaXT/decision_tree/decision_tree.pyx:
```diff
@@ -244,65 +244,79 @@ class DecisionTree(BaseModel):
+                        cur_node = cur_node.right_child
+                    depth += 1
+
+        all_idx = np.arange(0, X.shape[0], dtype=np.int32)
```
changing the all_idx to be all of X, means that you can not pass a indices, which has a smaller size than the X given. This might not be a desired use case, as one might want to pass in the entire X to avoid copying over.

NiklasPfister replied:
Good catch. I updated this now and also fixed this behavior at some other points. I also fixed that the sample weight are taken into account properly in the refit_leaf_nodes now.

svbrodersen commented on src/adaXT/random_forest/random_forest.py:
```diff
@@ -39,32 +38,52 @@ def tree_based_weights(
+                           replace=sampling_args['replace']), None)
     elif sampling == "honest_tree":
         indices = np.arange(0, n_rows)
         gen.shuffle(indices)
```
no need to do the shuffle if you decide to now use gen.choice later to get the fit_indices, pred_indices. However, in the case that replace is false, it might be faster to do the shuffle still instead of doing the random choice.

NiklasPfister replied:
This is needed so that for each tree we draw from different sets of indices for fitting and prediction. Its the only difference between honest_tree and honest_forest.

svbrodersen replied:
yes, but then in the case of doing sampling with "replace" there is no reason to do the random choice, no?
I mean the random shuffle would randomly shuffle all the indices, and then one could simply return first half as the fitting indices and the second half as prediction indices. It is only for when resampling is false, that we would need to still do the random choice again. With the current implementation we are adding another layer of randomness, which I don't see a reason for?

NiklasPfister replied:
Yes that is certainly true but adding an additional step might make the code more confusing (the difference between honest_tree and honest_forest becomes less clear then) and I don't think the computation is a big bottleneck here.

svbrodersen commented:
Added documentation to the predict_weights, predict_leaf and similarity of the decision tree and the random forest. Feel free to take a look, if you are happy with the version I have created or not. @NiklasPfister
